pkg/querier/distributor_queryable.go

```diff
@@ -38,46 +39,56 @@ type Distributor interface {
 	LabelValuesCardinality(ctx context.Context, labelNames []model.LabelName, matchers []*labels.Matcher) (uint64, *client.LabelValuesCardinalityResponse, error)
 }
 
-func newDistributorQueryable(distributor Distributor, iteratorFn chunkIteratorFunc, queryIngestersWithin time.Duration, logger log.Logger) QueryableWithFilter {
+func newDistributorQueryable(distributor Distributor, iteratorFn chunkIteratorFunc, cfgProvider distributorQueryableConfigProvider, logger log.Logger) QueryableWithFilter {
 	return distributorQueryable{
-		logger:               logger,
-		distributor:          distributor,
-		iteratorFn:           iteratorFn,
-		queryIngestersWithin: queryIngestersWithin,
+		logger:      logger,
+		distributor: distributor,
+		iteratorFn:  iteratorFn,
+		cfgProvider: cfgProvider,
 	}
 }
 
+type distributorQueryableConfigProvider interface {
+	QueryIngestersWithin(userID string) time.Duration
+}
+
 type distributorQueryable struct {
-	logger               log.Logger
-	distributor          Distributor
-	iteratorFn           chunkIteratorFunc
-	queryIngestersWithin time.Duration
+	logger      log.Logger
+	distributor Distributor
+	iteratorFn  chunkIteratorFunc
+	cfgProvider distributorQueryableConfigProvider
 }
 
 func (d distributorQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, err
+	}
 	return &distributorQuerier{
-		logger:               d.logger,
-		distributor:          d.distributor,
-		ctx:                  ctx,
-		mint:                 mint,
-		maxt:                 maxt,
-		chunkIterFn:          d.iteratorFn,
-		queryIngestersWithin: d.queryIngestersWithin,
+		logger:      d.logger,
+		distributor: d.distributor,
+		ctx:         ctx,
+		mint:        mint,
+		maxt:        maxt,
+		chunkIterFn: d.iteratorFn,
+		userID:      userID,
+		cfgProvider: d.cfgProvider,
 	}, nil
 }
 
-func (d distributorQueryable) UseQueryable(now time.Time, _, queryMaxT int64) bool {
+func (d distributorQueryable) UseQueryable(now time.Time, _, queryMaxT int64, userID string) bool {
 	// Include ingester only if maxt is within QueryIngestersWithin w.r.t. current time.
-	return d.queryIngestersWithin == 0 || queryMaxT >= util.TimeToMillis(now.Add(-d.queryIngestersWithin))
+	return d.cfgProvider.QueryIngestersWithin(userID) == 0 || queryMaxT >= util.TimeToMillis(now.Add(-d.cfgProvider.QueryIngestersWithin(userID)))
```

Comment:
`d.cfgProvider.QueryIngestersWithin(userID)` can change between calls, it would be best to get its value only once in this method.

Reply:
Using the same queryIngestersWithin value for all `distributorQuerier` calls as per https://github.com/grafana/mimir/pull/4287#discussion_r1120050033 in https://github.com/grafana/mimir/pull/4287/commits/a63a36cad58ee98abedbcf8d4ed2e1d8f6e69ad2